File: 0234-palindrome-linked-list/0234-palindrome-linked-list.cpp
/**
 *Definition for singly-linked list.
 *struct ListNode {
 *   int val;
 *   ListNode * next;
 *   ListNode() : val(0), next(nullptr) {}
 *   ListNode(int x) : val(x), next(nullptr) {}
 *   ListNode(int x, ListNode *next) : val(x), next(next) {}
 *};
 */
class Solution
{
    public:
        bool isPalindrome(ListNode *head)
        {
            ListNode *tempHead = head;
            stack<int> s;
            while (tempHead != NULL)
            {
                s.push(tempHead->val);
                tempHead = tempHead->next;
            }

            while (head != NULL)
            {
                if (head->val != s.top())
                {
                    return false;
                }
                s.pop();
                head = head->next;
            }
            return true;
        }
};

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseLL(ListNode head){
        if(head == null || head.next == null)   return head;
        
        ListNode newHead = reverseLL(head.next);
        ListNode tail = head.next;
        
        head.next = null;
        tail.next = head;
        
        return newHead;
    }
    public boolean isPalindrome(ListNode head) {
        if(head.next == null)   return true;
        
        int n = 0;
        ListNode prev = head;
        
        while(prev != null){
            prev = prev.next;
            ++n;
        }
        
        prev = null;
        ListNode slow = head;
        ListNode fast = head;
        ListNode head2 = null;
        
        while(fast != null && fast.next != null){
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }
        
        prev.next = null;
        
        head2 = slow;
        
        if(n % 2 != 0)  head2 = slow.next;
        
        head = reverseLL(head);
        
        while(head != null && head2 != null){
            if(head.val != head2.val)   return false;
            head = head.next;
            head2 = head2.next;
        }
        
        return true;
    }
}